File: core/topology.go
package core

import (
	"sync"
	"time"
)

// TopoNode represents a mesh node as seen in the global topology.
type TopoNode struct {
	ID          NodeID
	Caps        Capabilities
	Neighbors   []NodeID
	Epoch       uint64
	Seq         uint32
	Timestamp   int64
	Description string // free-form bio from the node's ANNOUNCE (key 8)
	Name        string // primary display label from the node's ANNOUNCE (key 9)
	Platform    string // OS/device string from the node's ANNOUNCE (key 10)
	PublicKey   []byte // 32-byte Ed25519 key from ANNOUNCE (key 6); used for chat encryption
	LastSeen    time.Time
}

// Topology maintains the global mesh graph learned via ANNOUNCE packets.
type Topology struct {
	mu     sync.RWMutex
	nodes  map[NodeID]*TopoNode
	expiry time.Duration
}

func NewTopology() *Topology {
	t := &Topology{
		nodes:  make(map[NodeID]*TopoNode),
		expiry: 90 * time.Second,
	}
	go t.reap()
	return t
}

// Update inserts or refreshes a topology node, ignoring stale epoch/seq updates.
func (t *Topology) Update(n TopoNode) bool {
	t.mu.Lock()
	defer t.mu.Unlock()
	existing, ok := t.nodes[n.ID]
	if ok && (existing.Epoch > n.Epoch || (existing.Epoch == n.Epoch && existing.Seq >= n.Seq)) {
		return false
	}
	n.LastSeen = time.Now()
	t.nodes[n.ID] = &n
	return true
}

// BFSPath returns the source-route path from `from` to `to` (excluding `from`, including `to`).
// Returns nil if no path found.
func (t *Topology) BFSPath(from, to NodeID) []NodeID {
	return t.BFSPathFromSource(from, to, nil)
}

// BFSPathFromSource is BFSPath but seeds `from`'s adjacency with fromNeighbors. A node never
// appears in its own topology (it doesn't process its own ANNOUNCE), so a plain BFS from LocalID
// dead-ends immediately — every route would be nil and every unicast would silently flood. Passing
// the local node's direct neighbor IDs here lets the search take its first hop and find real
// multi-hop source routes (used by Router.SelectRoute with the local NeighborTable).
func (t *Topology) BFSPathFromSource(from, to NodeID, fromNeighbors []NodeID) []NodeID {
	t.mu.RLock()
	defer t.mu.RUnlock()

	if from == to {
		return []NodeID{}
	}

	visited := map[NodeID]bool{from: true}
	prev := map[NodeID]NodeID{}
	queue := []NodeID{from}

	for len(queue) > 0 {
		cur := queue[0]
		queue = queue[1:]
		if cur == to {
			// Reconstruct path (excluding 'from', including 'to')
			var path []NodeID
			for cur != from {
				path = append([]NodeID{cur}, path...)
				cur = prev[cur]
			}
			return path
		}
		// Adjacency: topology entry's neighbors, plus the seeded neighbors for the source node.
		var neighbors []NodeID
		if node, ok := t.nodes[cur]; ok {
			neighbors = node.Neighbors
		}
		if cur == from {
			neighbors = append(append([]NodeID(nil), neighbors...), fromNeighbors...)
		}
		for _, nb := range neighbors {
			if !visited[nb] {
				visited[nb] = true
				prev[nb] = cur
				queue = append(queue, nb)
			}
		}
	}
	return nil
}

// Nodes returns a snapshot of all known topology nodes.
func (t *Topology) Nodes() []TopoNode {
	t.mu.RLock()
	defer t.mu.RUnlock()
	out := make([]TopoNode, 0, len(t.nodes))
	for _, n := range t.nodes {
		out = append(out, *n)
	}
	return out
}

// GetNode returns a topology node by ID.
func (t *Topology) GetNode(id NodeID) (TopoNode, bool) {
	t.mu.RLock()
	defer t.mu.RUnlock()
	n, ok := t.nodes[id]
	if !ok {
		return TopoNode{}, false
	}
	return *n, true
}

// ExpireNode forcibly removes a node — useful for testing.
func (t *Topology) ExpireNode(id NodeID) {
	t.mu.Lock()
	defer t.mu.Unlock()
	delete(t.nodes, id)
}

func (t *Topology) reap() {
	ticker := time.NewTicker(15 * time.Second)
	for range ticker.C {
		t.Reap()
	}
}

// Reap removes topology entries that have exceeded the expiry duration.
// Exported for testing.
func (t *Topology) Reap() {
	t.mu.Lock()
	now := time.Now()
	for id, n := range t.nodes {
		if now.Sub(n.LastSeen) > t.expiry {
			delete(t.nodes, id)
		}
	}
	t.mu.Unlock()
}

File: android/bleedge-protocol/src/main/java/cz/arnal/bleedge/protocol/Topology.kt
package cz.arnal.bleedge.protocol

import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.concurrent.read
import kotlin.concurrent.write

/**
 * A node learned from a verified signed announce (§8.5, §12). Freshness is keyed
 * on (epoch, seq); local expiry uses [receivedAtMs], not the signed timestamp.
 */
data class TopoNode(
    val id: NodeId,
    val publicKey: ByteArray,
    val caps: Capabilities,
    val neighbors: List<NodeId>,
    val epoch: Long,
    val seq: Long,
    val timestamp: Long,
    val name: String = "",
    val description: String = "",
    val platform: String = "",
    val receivedAtMs: Long = System.currentTimeMillis(),
)

/**
 * Thread-safe global mesh topology graph built from verified announces. Stale
 * announces (lower epoch, or equal epoch with non-increasing seq) are rejected.
 */
class Topology(private val expiryMs: Long = BLEEdge.TOPOLOGY_TIMEOUT_MS) {
    private val lock = ReentrantReadWriteLock()
    private val nodes = mutableMapOf<String, TopoNode>()

    /** Returns true only when [node] is strictly newer and was therefore stored (§8.5). */
    fun update(node: TopoNode): Boolean = lock.write {
        val existing = nodes[node.id.toHex()]
        if (existing != null && !isNewer(node.epoch, node.seq, existing.epoch, existing.seq)) return@write false
        nodes[node.id.toHex()] = node.copy(receivedAtMs = System.currentTimeMillis())
        true
    }

    private fun isNewer(epoch: Long, seq: Long, curEpoch: Long, curSeq: Long): Boolean =
        epoch > curEpoch || (epoch == curEpoch && seq > curSeq)

    fun getNode(id: NodeId): TopoNode? = lock.read { nodes[id.toHex()] }
    fun allNodes(): List<TopoNode> = lock.read { nodes.values.toList() }
    fun remove(id: NodeId) = lock.write { nodes.remove(id.toHex()) }

    fun publicKeyFor(id: NodeId): ByteArray? =
        lock.read { nodes[id.toHex()]?.publicKey?.takeIf { it.size == BLEEdge.PUBLIC_KEY_BYTES } }

    /**
     * BFS shortest path over the learned graph from [from] to [to] (§10.4).
     * Returns the hops excluding [from] and including [to], or null if no path.
     */
    fun bfsPath(from: NodeId, to: NodeId): List<NodeId>? = bfsPathFromSource(from, to, emptyList())

    /**
     * [bfsPath] but seeds [from]'s adjacency with [fromNeighbors]. A node never appears in its own
     * topology (it doesn't process its own ANNOUNCE), so a plain BFS from the local id dead-ends
     * immediately — every route would be null and every unicast would silently flood. Passing the
     * local node's direct neighbor ids lets the search take its first hop and find real multi-hop
     * source routes (used by [Router.selectRoute] with the local neighbor table).
     */
    fun bfsPathFromSource(from: NodeId, to: NodeId, fromNeighbors: List<NodeId>): List<NodeId>? = lock.read {
        if (from == to) return@read emptyList()
        val visited = mutableSetOf(from.toHex())
        val prev = mutableMapOf<String, NodeId>()
        val queue = ArrayDeque<NodeId>()
        queue.add(from)
        var found = false
        while (queue.isNotEmpty()) {
            val cur = queue.removeFirst()
            if (cur == to) { found = true; break }
            val neighbors = (nodes[cur.toHex()]?.neighbors ?: emptyList()) +
                (if (cur == from) fromNeighbors else emptyList())
            for (nb in neighbors) {
                if (visited.add(nb.toHex())) { prev[nb.toHex()] = cur; queue.add(nb) }
            }
        }
        if (!found) return@read null
        val path = ArrayDeque<NodeId>()
        var cur = to
        while (cur != from) {
            path.addFirst(cur)
            cur = prev[cur.toHex()] ?: return@read null
        }
        path.toList()
    }

    fun reap() = lock.write {
        val now = System.currentTimeMillis()
        nodes.entries.filter { now - it.value.receivedAtMs > expiryMs }.forEach { nodes.remove(it.key) }
    }
}
